### ISSUE TITLE
Unable to catch RedisClient\Exception\EmptyResponseException

### ISSUE THREAD

**Issue by SmoshySmosh** [cheprasov/php-redis-client#56]
I fully admit that I may be doing this wrong, but the call stack is not being very helpful in terms of where I need to catch this connection error.

Here is what I am doing, please let me know if you see something out of the ordinary.

Using Pimple:
```php
$container['redis'] = function (Container $c) {
    $settings = $c['app.settings']['redis'];

    try {
        $redis = ClientFactory::create([
            'server'  => \sprintf('%s:%s', $settings['host'], $settings['port']),
            'timeout' => $settings['timeout'],
        ]);

        return $redis;
    } catch (EmptyResponseException $e) {
        $c['app.logger']->error('Failed to connect to redis... Retrying...');
       
        // Try to reconnect
        return $container['redis']($c);
    }
};
```

Full Stack trace from php:
```
Fatal error: Uncaught RedisClient\Exception\EmptyResponseException: Empty response. Please, check connection timeout. in /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Protocol/RedisProtocol.php:111
Stack trace:
#0 /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Protocol/RedisProtocol.php(164): RedisClient\Protocol\RedisProtocol->read()
#1 /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Protocol/RedisProtocol.php(171): RedisClient\Protocol\RedisProtocol->sendRaw('*5\r\n$7\r\nEVALSHA...')
#2 /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Client/AbstractRedisClient.php(174): RedisClient\Protocol\RedisProtocol->send(Array)
#3 /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Client/AbstractRedisClient.php(155): RedisClient\Client\AbstractRedisClient->executeProtocolCommand(Object(RedisClient\Protocol\RedisProtocol), Array, Array)
#4 /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Client/AbstractRedisClient.php(142): RedisClient\Client\AbstractRedisClient->exec in /var/www/vendor/cheprasov/php-redis-client/src/RedisClient/Protocol/RedisProtocol.php on line 111
```

Note: It only throws the exception every once in a while. It could be something AWS related, but I'd still like to be able to catch the error and retry the connection.

**Comment by cheprasov** [comment]
Hello @SmoshySmosh 

I see that you wrap it `try-catch` a creation of new instance, but, for catching `EmptyResponseException` exception, you should to wrap a client method, that may throws an exception.

By the trace, I see you use `$Redis->evalsha(...);` therefore, you should to use try-catch something like that, for example:

```php
    try {
        return $Redis->evalsha($script);
    } catch (EmptyResponseException $e) {
        // some action
    }
```

**Comment by SmoshySmosh** [comment]
Hi @cheprasov 

I am actually using this library with your [php-redis-lock](https://github.com/cheprasov/php-redis-lock) library, so I'm not calling `$redis->evalsha(...);` directly. Here is an example.

```php
// Obtain Lock
// Obtain Lock
$state->lock = new RedisLock($this->redis, $jobId, RedisLock::FLAG_CATCH_EXCEPTIONS);

// Log starting to aquire lock
$this->logger->info('Acquiring lock...');

// Aquire lock
if (!$state->lock->acquire(1200, 1)) {
    $this->logger->info('Failed to acquire lock after 1 second...');

    // Unset Lock
    unset($state->lock);

    // Cycle Job
    if ($this->queue->cycle($job)) {
        $this->logger->info('Job was successfully cycled...');
    } else {
        $this->logger->error('Job failed to cycle...');
    }

    return false;
}
```

In theory, `RedisLock::FLAG_CATCH_EXCEPTIONS` should be catching the exception and returning a false in the if statement, but instead i'm getting a Unhandled Exception error which I am trying to catch.

Additionally, I have found that when you use a redis version number in your configuration, the returned object does not implement the proper interface, so it can not be passed to php-redis-lock. A type passed error is thrown.

**Comment by cheprasov** [comment]
Hey @SmoshySmosh 

> Additionally, I have found that when you use a redis version number in your configuration, the returned object does not implement the proper interface, so it can not be passed to php-redis-lock. A type passed error is thrown.

Thanks a lot for report the bug, I fixed it in version 1.7.2 of the RedisClient.

Also, `RedisLock::FLAG_CATCH_EXCEPTIONS` means that RedisLock class will not throw exceptions , but it will not catch exceptions from used libraries. Sorry for ambiguous of flag name, I will create other flag for it in the future version like `RedisLock::FLAG_DO_NOT_THROW_EXCEPTIONS` instead of `RedisLock::FLAG_CATCH_EXCEPTIONS`
